tasks.findByName magically *creates* tasks, preventing future mutation (!!)

after hours of debugging the entirely unhelpful message:
```
Caused by: java.lang.IllegalStateException: Cannot add Mutate rule 
'org.gradle.language.base.plugins.ComponentModelBasePlugin$Rules#closeSourcesForBinaries(org.gradle.platform.base.BinaryContainer, org.gradle.language.base.ProjectSourceSet)'
for model element 'binaries' when element is in state GraphClosed.
```

the issue is the use of findByName('assemble').  This cause Gradle to *create* the assembly task, preventing modification of say the binaries later.

Pulpit:
This is why I hate things that have too much magic.  Only a line by line binary search caused to me find this, since the stack trace is thrown much later when the project's tasks have to be resolved, and nowhere near either the mutation or the findByName.  SHEESH!

@brunobowden I'm fixing this to allow native dependencies, hopefully you can rollback your workaround for plugin syntax once I merge this in.

This'll be something else I ask about tomorrow.

On Thu, Jun 11, 2015 at 10:41 PM Advay Mengle <notifications@github.com>
wrote:

> after hours of debugging the entirely unhelpful message:
>
> Caused by: java.lang.IllegalStateException: Cannot add Mutate rule 'org.gradle.language.base.plugins.ComponentModelBasePlugin$Rules#closeSourcesForBinaries(org.gradle.platform.base.BinaryContainer, org.gradle.language.base.ProjectSourceSet)' for model element 'binaries' when element is in state GraphClosed.
>
> the issue is the use of findByName('assemble'). This cause Gradle to
> *create* the assembly task, preventing modification of say the binaries
> later.
>
> Pulpit:
> This is why I hate things that have too much magic. Only a line by line
> binary search caused to me find this. Since the stack trace is thrown much
> later when the project's tasks have to be resolved.
>
> @brunobowden <https://github.com/brunobowden> I'm fixing this to allow
> native dependencies, hopefully you can rollback your workaround for plugin
> syntax once I merge this in.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/j2objc-contrib/j2objc-gradle/issues/156>.
>

Wow, that's an amazing find Advay

On Thu, Jun 11, 2015 at 10:43 PM Bruno <brunob@gmail.com> wrote:

> This'll be something else I ask about tomorrow.
>
> On Thu, Jun 11, 2015 at 10:41 PM Advay Mengle <notifications@github.com>
> wrote:
>
>> after hours of debugging the entirely unhelpful message:
>>
>> Caused by: java.lang.IllegalStateException: Cannot add Mutate rule 'org.gradle.language.base.plugins.ComponentModelBasePlugin$Rules#closeSourcesForBinaries(org.gradle.platform.base.BinaryContainer, org.gradle.language.base.ProjectSourceSet)' for model element 'binaries' when element is in state GraphClosed.
>>
>> the issue is the use of findByName('assemble'). This cause Gradle to
>> *create* the assembly task, preventing modification of say the binaries
>> later.
>>
>> Pulpit:
>> This is why I hate things that have too much magic. Only a line by line
>> binary search caused to me find this. Since the stack trace is thrown much
>> later when the project's tasks have to be resolved.
>>
>> @brunobowden <https://github.com/brunobowden> I'm fixing this to allow
>> native dependencies, hopefully you can rollback your workaround for plugin
>> syntax once I merge this in.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/j2objc-contrib/j2objc-gradle/issues/156>.
>>
>

--- a/src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy
+++ b/src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy
@@ -186,14 +186,19 @@ class J2objcPlugin implements Plugin<Project> {
     // The before task must already exist.
     private def lateDependsOn(Project proj, String afterTaskName, String beforeTaskName) {
         assert null != proj.tasks.findByName(beforeTaskName)
-        def afterTask = proj.tasks.findByName(afterTaskName)
-        if (afterTask != null) {
-            afterTask.dependsOn beforeTaskName
-        } else {
-            proj.tasks.whenTaskAdded { task ->
-                if (task.name == afterTaskName) {
-                    task.dependsOn beforeTaskName
-                }
+        // You can't just call tasks.findByName on afterTaskName - for certain tasks like 'assemble' for
+        // reasons unknown, the Java plugin creates - right there! - the task; this prevents
+        // later code from modifying binaries, sourceSets, etc.  If you see an error
+        // mentioning 'state GraphClosed' saying you can't mutate some object, see if you are magically
+        // causing Gradle to make the task by using findByName!  Issue #156
+
+        // tasks.all cleanly calls this closure on any existing elements and for all elements
+        // added in the future.
+        // TODO: Find a better way to have afterTask depend on beforeTask, without
+        // materializing afterTask early.
+        proj.tasks.all { task ->
+            if (task.name == afterTaskName) {
+                task.dependsOn beforeTaskName
             }
         }
     }

